### PR TITLE
Deactivation cleanup

### PR DESCRIPTION
**Pull request by arunagw**
Remove hound from repo and org when user deactivate repo.

### CODE REVIEW

**Comment by salbertson** [comment]
@arunagw, what happens when Hound is enabled for multiple repos in an organization? Should Hound only be removed from an organization if no other repos are being reviewed?

File: app/services/manage_hound.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  SERVICES_TEAM_NAME = "Services"
+
+  pattr_initialize :repo_name, :github
+
+  def self.run(repo_name, github)
+    new(repo_name, github).run
+  end
+
+  def run
+    raise "Must be implemented in child class"
```
Comment (gylaz) [inline app/services/manage_hound.rb]: What do you think about doing the same as in `style_guide/base.rb` class?
```ruby
raise NotImplementedError.new("must implement ##{__method__}")
```

File: app/services/repo_activator.rb [hunk 2]
```diff
@@ -24,6 +22,12 @@ def activate_repo
     end
   end
 
+  def deactivate_repo
+    change_repository_state_quietly do
+      remove_hound_from_repo && delete_webhook && repo.deactivate
+    end
```
Comment (gylaz) [inline app/services/repo_activator.rb]: Why not just have this block in `deactivate`?

File: app/services/remove_hound_from_repo.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      team_id = team.id
+      remove_repo_from_team(team_id)
+      delete_team(team_id)
+    end
+  end
+
+  def delete_team(team_id)
+    team_repos = github.team_repos(team_id)
+
+    if team_repos.count == 0
```
Comment (gylaz) [inline app/services/remove_hound_from_repo.rb]: `team_repos.empty?`

File: lib/github_api.rb [hunk 4]
```diff
@@ -150,12 +165,14 @@ def user_repos
     authorized_repos(client.repos)
   end
 
-  def org_repos
-    repos = orgs.flat_map do |org|
-      client.org_repos(org[:login])
+  def all_org_repos
```
Comment (gylaz) [inline lib/github_api.rb]: What do you think about naming this `repos_from_all_orgs`? A bit more descriptive, as `all_org_repos` reads as in "all repos from org".

**Comment by gylaz** [comment]
Thoughts on not removing the "Services" team? What is the benefit of removing it?

**Comment by arunagw** [comment]
@gylaz @salbertson please have a look again. 

cc/ @samphippen

File: spec/lib/github_api_spec.rb [hunk 5]
```diff
@@ -328,4 +321,73 @@
...
+        team_id, username, token
+      )
+
+      api.remove_user_from_team(team_id, username)
+
+      expect(request).to have_been_requested
+    end
+  end
+
+  private
```
Comment (teoljungberg) [inline spec/lib/github_api_spec.rb]: Our styleguide says that we should avoid having `private` in specs

File: app/services/manage_hound.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  SERVICES_TEAM_NAME = "Services"
+
+  pattr_initialize :repo_name, :github
+
+  def self.run(repo_name, github)
+    new(repo_name, github).run
+  end
+
+  def run
+    raise NotImplementedError.new("must implement ##{__method__}")
```
Comment (teoljungberg) [inline app/services/manage_hound.rb]: How about writing
```ruby
raise NotImplementedError, "must implement ##{__method__}"
```
Instead of instantiating the error class with the message?

File: app/services/remove_hound_from_repo.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+  def remove_hound_from_repo
+    github.remove_collaborator(repo_name, github_username)
+  end
+
+  def remove_services_team_from_repo
+    team = find_services_team
+
+    if team
+      team_id = team.id
```
Comment (teoljungberg) [inline app/services/remove_hound_from_repo.rb]: How about inlining this local variable instead?

**Comment by arunagw** [comment]
This PR is in WIP again. Will ping back once ready to review

**Comment by arunagw** [comment]
@gylaz I think we should not remove team "Services" as user or any other services might have added something to it and we don't want to mess with that.

Remove repo and user from that team seems fine.

File: app/models/github_team.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
+class GithubTeam
+  def initialize(team, github)
```
Comment (teoljungberg) [inline app/models/github_team.rb]: We have `attr_extras` in the project, how about using `pattr_initialize here?

File: app/services/manage_hound.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  private
+
+  def repo
+    @repo ||= github.repo(repo_name)
+  end
+
+  def github_username
+    @github_username ||= ENV.fetch("HOUND_GITHUB_USERNAME")
```
Comment (teoljungberg) [inline app/services/manage_hound.rb]: Maybe this does not need to be memoized?

File: app/services/manage_hound.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
+require "app/models/github_team"
+
+class ManageHound
+  SERVICES_TEAM_NAME = "Services"
```
Comment (teoljungberg) [inline app/services/manage_hound.rb]: Maybe this can be renamed from `GITHUB_TEAM_NAME`?

**Comment by teoljungberg** [comment]
Few comments, and fixed a few of them. But besides that, LGTM

File: spec/lib/github_api_spec.rb (hunk 5 above)
Comment (gylaz) [inline spec/lib/github_api_spec.rb]: It does? Mind pasting the rule, or linking to it?

Comment (teoljungberg) [inline spec/lib/github_api_spec.rb]: https://github.com/thoughtbot/guides/tree/master/style/testing, First
bullet-point

File: app/services/repo_activator.rb (hunk 2 above)
Comment (arunagw) [inline app/services/repo_activator.rb]: I think it's good to keep the same as `activate` if you look into this file it's now same as we are doing in `activate`

File: app/models/github_team.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  def has_pull_permission?
+    team.permission == "pull"
+  end
+
+  def remove_user_from_team(github_username)
+    team_repos = github.team_repos(team.id)
+
+    if team_repos.empty?
```
Comment (arunagw) [inline app/models/github_team.rb]: Should we have this check here or in `RemoveHoundFromRepo` class?

Comment (gylaz) [inline app/models/github_team.rb]: Seems good here.

**Comment by arunagw** [comment]
This one looks good to merge? 

What we are doing here is in case of org repo.
1. Removing repo from "Services" team.
2. Removing Hound User from team if no more repo in "Services" team

File: app/models/github_team.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
+class GithubTeam
+  pattr_initialize :team, :github
+
+  def add_repo(repo_name)
+    ensure_push_permission
+    add_repo_to_team(repo_name)
+  end
+
+  def remove_repo(repo_name)
+    remove_repo_from_team(repo_name)
```
Comment (gylaz) [inline app/models/github_team.rb]: I'm curious about this coding approach -- `remove_repo_from_team` contains just one line. Why abstract it to yet another method?

File: app/models/github_team.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  def remove_user(github_username)
+    remove_user_from_team(github_username)
+  end
+
+  def add_user(github_username)
+    add_user_to_team(github_username)
```
Comment (gylaz) [inline app/models/github_team.rb]: Same question here. This approach generates more code, without an apparent benefit. `github.add_repo_to_team(team.id, repo_name)` is pretty readable to me. Am I missing something?

File: app/services/manage_hound.rb [hunk 14]
```diff
@@ -0,0 +1,43 @@
+require "app/models/github_team"
```
Comment (gylaz) [inline app/services/manage_hound.rb]: Why is this necessary?

File: app/services/manage_hound.rb [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+  def github_username
+    ENV.fetch("HOUND_GITHUB_USERNAME")
+  end
+
+  def decorated_services_team
+    if find_services_team
+      GithubTeam.new(find_services_team, github)
+    else
+      nil
```
Comment (gylaz) [inline app/services/manage_hound.rb]: `else` isn't necessary here, technically.

File: app/services/remove_hound_from_repo.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
+require "app/services/manage_hound"
```
Comment (gylaz) [inline app/services/remove_hound_from_repo.rb]: Why is this necessary?

Comment (gylaz) [inline app/services/remove_hound_from_repo.rb]: If to make the tests happy, then this `require` should be in the test, no?

File: app/services/remove_hound_from_repo.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    end
+  end
+
+  private
+
+  def remove_hound_from_repo
+    github.remove_collaborator(repo_name, github_username)
+  end
+
+  def remove_services_team_from_repo
```
Comment (gylaz) [inline app/services/remove_hound_from_repo.rb]: Should this be `remove_repo_from_services_team`?

File: app/models/github_team.rb (hunk 11 above)
Comment (gylaz) [inline app/models/github_team.rb]: Looking at this a bit more, I'm torn. On one hand we want to use the "Tell don't ask" principal, on the other hand, reading [this line of code](https://github.com/thoughtbot/hound/pull/603/files#diff-153e44c17166a5b3c6f96a2a1988860eR21) it would be surprising to me if that didn't remove the user from the team.

File: spec/features/repo_list_spec.rb [hunk 18]
```diff
@@ -150,4 +155,8 @@ def with_job_delay
   ensure
     Resque.inline = true
   end
+
+  def username
+    "houndci"
+  end
```
Comment (gylaz) [inline spec/features/repo_list_spec.rb]: Should this be a `let`?

File: spec/lib/github_api_spec.rb [hunk 19]
```diff
@@ -114,7 +108,7 @@
     it "returns changed files in a pull request" do
       hound_token = ENV["HOUND_GITHUB_TOKEN"]
       api = GithubApi.new(hound_token)
-      pull_request = double(:pull_request, full_repo_name: "thoughtbot/hound")
+      pull_request = double(:pull_request, full_repo_name: full_repo_name)
```
Comment (gylaz) [inline spec/lib/github_api_spec.rb]: Please use the string version of double names. `double("PullRequest"...`

File: spec/lib/github_api_spec.rb [hunk 20]
```diff
@@ -133,15 +127,14 @@
...
     it "adds comment to GitHub pull request" do
       api = GithubApi.new("authtoken")
-      repo_name = "test/repo"
       pull_request_number = 2
       comment = "test comment"
       commit_sha = "commitsha"
       file = "test.rb"
       patch_position = 123
-      commit = double(:commit, repo_name: repo_name, sha: commit_sha)
+      commit = double(:commit, repo_name: full_repo_name, sha: commit_sha)
```
Comment (gylaz) [inline spec/lib/github_api_spec.rb]: `double("Commit"...`

File: spec/lib/github_api_spec.rb (hunk 19 above)
Comment (gylaz) [inline spec/lib/github_api_spec.rb]: Since we are touching this line, can we make that update as well? It's a nice way to prep ourselves for RSpec 3.1 upgrade.

File: spec/models/github_team_spec.rb [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+  end
+
+  describe "#add_user" do
+    it "calls github with the team id and passed username" do
+      id = 1234
+      team = double("OctoKitTeam", id: id, permission: non_pull_permission)
+      github = double(
+        "Github",
+        add_user_to_team: true
+      )
```
Comment (gylaz) [inline spec/models/github_team_spec.rb]: Seems like this could fit on one line.

File: spec/models/github_team_spec.rb [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+    "pull"
+  end
+
+  def repo_name
+    "foo/bar"
+  end
+
+  def username
+    "houndci"
+  end
```
Comment (gylaz) [inline spec/models/github_team_spec.rb]: These all look like good candidates for `let`s.

**Comment by gylaz** [comment]
Left a few comments/suggestions. Looks good to me. Merge at your discretion.

File: spec/models/github_team_spec.rb (hunk 22 above)
Comment (arunagw) [inline spec/models/github_team_spec.rb]: It seems that we don't want to use `let` here  https://github.com/thoughtbot/guides/tree/master/best-practices#testing

I have seen that we have used `let` in `builds_spec`. 

What do you think? Should we add?

Comment (gylaz) [inline spec/models/github_team_spec.rb]: We don't want to use `let`, because it has higher-level code issues. We shouldn't just re-implement `let`s because of that rule.
